### PR TITLE
perf(travel): make the pathfinding A* actually search towards the goal

### PR DESCRIPTION
**Pull request by MathiasGruber**
The travel map's A* was not doing any A*. Its heuristic measured how far a tile is from where the walk **started**:

```ts
estimateFromNodeToGoal: (tile) => this.grid.distance(tile, origin),
```

`h(n)` therefore grew in step with `g(n)` instead of shrinking towards the target, `f = g + h` stayed monotonic in `g`, and the search flooded outwards in every direction — every reachable tile inside the goal's cost radius, each expansion allocating a Grid, an array and a filtered array. It was *worse than passing no heuristic at all*: identical search work, plus a `distance()` call per relaxed node.

## Aiming it at the goal

The scale matters as much as the direction. A step costs `to.cost + from.cost`, so it is never cheaper than twice the cheapest passable tile on that grid — and this one class serves grids with different floors: combat sets every tile to `1` (min step 2), sector and window grids use `walkCost + 1` (min step 4). A hardcoded factor cannot serve both: `2` is a no-op on the travel grids, and `4` overestimates in combat, which would bend AI movement and barrier lookups. So the bound is read off the grid once, in the constructor, alongside the path cache that already assumes tile state holds still.

## Why the search itself was replaced — in the same commit

Fixing the heuristic made it actually reorder the frontier — which turned a latent bug in `abstract-astar` into wrong answers. Its `MinHeap` sifts **down** when a key decreases (a decrease has to sift **up**), and `removeMinimum` leaves the moved item's stale index behind in its index map. Both corrupt the pop order, so a tile gets settled before its cheapest route is known:

> 4 of 68 paths on a sector-sized grid with mixed terrain costs came back ~1% longer than optimal.

That is not just a longer walk. `getBarriersBetween` derives *which barriers sit between attacker and target* from this path, so a different route is different damage absorbed.

The replacement is A* over a lazy-deletion heap: a cheaper route pushes a second entry rather than moving the existing one, and whichever entry surfaces after a tile is settled is skipped. Never decreasing a key sidesteps both bugs. It is also slightly faster than the library, so the correctness costs nothing.

These started as two commits and were squashed into one, because the heuristic fix *without* the new search is worse than either side of it — 15 of 1150 paths wrong on a 26×26 grid, 54 of 2814 on a 40×40 one, worst case 8.3% longer, where the code before the branch was 0 for 0. This repo rebase-merges, so splitting them would leave that state in `main` for anyone bisecting through it.

## Measured

This repo's own grid shapes, the same origin/target pairs throughout, a fresh calculator each run so the cache never answers:

| | before | heuristic fixed | + own search |
|---|---|---|---|
| 78×78 window, mixed terrain, 12% blocked | 6.33 ms | 1.62 ms | **1.49 ms** |
| 13×13 combat arena | 0.223 ms | 0.044 ms | **0.044 ms** |

Per path, averaged over 20 random reachable pairs.

Two things about where this actually lands, since the finding that prompted the PR overstates them and I would rather the record be right:

**It is not in a render loop.** All four `getShortestPath` call sites in `Sector.tsx` are event handlers — the click-to-move handler, the post-move remainder walk, the border-crossing walk, and the websocket handler for other players' movement. `Sector.tsx:1638` is a scene-disposal `useEffect` cleanup, and `libs/threejs/sector.ts` never calls it at all. So this is not dropped frames while panning; it is latency on an interaction.

**A busy sector was never the tens-of-milliseconds case.** Every other player's step does cost a fresh search — the cache is keyed on origin+target, so it never hits — but a step is one tile, and even an uninformed Dijkstra stops as soon as it pops an adjacent target. Measured on the 78×78 window grid:

| distance | before | after |
|---|---|---|
| 1 tile (a broadcast step) | 0.098 ms | 0.029 ms |
| 2 tiles | 0.104 ms | 0.026 ms |
| 5 tiles | 0.581 ms | 0.120 ms |
| 10 tiles | 1.565 ms | 0.456 ms |

So the busy-sector case gets ~3.4× cheaper but was already sub-millisecond. The tens of milliseconds were only ever in the long queries — clicking a far tile, or the cross-border nav on the unified window grid — which is exactly where this pays off.

**313 random pairs** across combat, window and maze-shaped grids: every path matches an independent Dijkstra's cost exactly.

## Tests

`app/tests/libs/hexgrid.astar.test.ts` compares against a deliberately naive Dijkstra written in the test rather than against stored paths, so it fails on any heuristic that overestimates. Verified it has teeth: forcing the scale from 2 to 4 fails five of them, including a concrete 109-vs-107 path cost.

It also pins the parts that are easy to get subtly wrong later — that the bound follows the cheapest *passable* tile (a cheap blocked tile must not drag it down), that a grid with no passable tile or a nonsense cost falls back to a plain Dijkstra rather than to a broken heuristic, and that a costly-but-passable tile (`ai_v2` marks occupied tiles cost 100) is walked around rather than through.

## The behaviour change worth knowing about

A search that is actually guided picks a **different one of the equally cheap routes**. Not a longer one — the same cost, and on a uniform grid the same tile count — but different tiles. On the 12×10 battle grid, across all 14,280 ordered pairs:

| | |
|---|---|
| routes that differ | **9,844 (68.9%)** |
| paths whose cost differs | 0 |
| paths whose tile count differs | 0 |

Two consumers read the route rather than its cost, and both are server-side combat:

- `getBarriersBetween` derives *which barriers sit between attacker and target*, so a barrier on a given tile changes its verdict for **4.52%** of (pair, tile) combinations.
- `ai_v2`'s `move_towards_opponent` steps onto `path[1]`, which differs for **19%** of pairs — so an AI's first step changes that often.

Battles are persisted, so this lands inside fights that are already running. `mapDistancesToTarget` is unaffected: it keys on `path.length`, which does not move on the combat grid (0 of 14,280 pairs).

It is not a regression — the old route was itself an arbitrary pick among ties, whatever Dijkstra's pop order happened to yield, and there is no way to speed the search up without moving that choice. But it is a real, live change to combat damage and it should be a deliberate merge rather than a surprise.

`app/tests/libs/hexgrid.astar.test.ts` now pins one route as a tripwire, since cost-only assertions cannot see this: reversing the neighbour order keeps every path optimal, moves 77.5% of routes, and leaves the suite green.

## Risk

- The route change above.
- `abstract-astar` is now unused by this file but stays in `package.json` — dropping it means a lockfile change, and the local bun rewrites that file wholesale (it cannot parse `lockfileVersion: 3`), so it belongs in its own PR.

---

> [!NOTE]
> **Medium Risk**
> Correctness improves for path cost, but many tie-breaking routes change and can alter live combat barrier/damage via `getBarriersBetween`; `minStepCost` must be rebuilt if tile costs decrease after construction.
> 
> **Overview**
> **Fixes sector/travel pathfinding** by pointing A* at the goal and swapping out the `abstract-astar` dependency for an in-house lazy-deletion heap implementation.
> 
> `PathCalculator` no longer uses `distance(tile, origin)` as the heuristic (which behaved like uninformed search). It snapshots **`minStepCost`** from the cheapest passable tile and uses `minStepCost * distance(tile, target)` so combat and sector/window grids both get a valid, non-overestimating estimate. **`findShortestPath`** implements A* with push-on-decrease heap entries to avoid `abstract-astar` MinHeap bugs that could return ~1% longer routes once the heuristic actually reordered the frontier.
> 
> Docs and tests call out that **only path cost is guaranteed**—among equal-cost routes, which tiles are returned can change and affects **`getBarriersBetween`** (combat barrier absorption). New **`hexgrid.astar.test.ts`** compares costs to a naive Dijkstra, covers `minStepCost` edge cases, and pins one tie-break route as a deliberate tripwire.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7c2a0aa538bfe11bd8143f235c023fe669ddfda6. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: app/tests/libs/hexgrid.astar.test.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+});
+
+describe("PathCalculator.getShortestPath", () => {
+  it("returns a contiguous path from origin to target", () => {
+    const grid = makeGrid(9, 9, () => 2);
+    const path = new PathCalculator(grid).getShortestPath(at(grid, 0, 0), at(grid, 8, 8));
+    expect(path?.[0]).toBe(at(grid, 0, 0));
+    expect(path?.[path.length - 1]).toBe(at(grid, 8, 8));
+    path?.slice(1).forEach((tile, index) => {
+      expect(grid.distance(path[index] as TerrainHex, tile)).toBe(1);
```
[the remark this answers is not in the document]
Comment (MathiasGruber) [reply app/tests/libs/hexgrid.astar.test.ts]: Fixed in a8a9d44c0. The typing was the smaller half of it — the real problem was that the contiguity walk sat behind `path?.slice(1)`, so an implementation returning `undefined` skipped every step of the loop and the test passed without checking anything. It now narrows through a `found()` helper first; with `getShortestPath` forced to return `undefined` the file goes from 14 pass to 6 failures instead of quietly staying green.
